### PR TITLE
Annotate dead code lint with notes about ignored derived impls

### LINKED ISSUE #92726: Annotate dead-code lint with a note that Debug impls are ignored
PR https://github.com/rust-lang/rust/pull/85200 modified the dead-code lint so that it ignores Debug impls (and perhaps a few others, I forget) for the purposes of dead-code detection. This is pretty reasonable, but it does create false dead-code warnings when one has structs that exist solely to be debug printed. As evidenced by #88900, users are confused by this in practice.

@pmetzger had a [good suggestion](https://github.com/rust-lang/rust/issues/88900#issuecomment-991634289):

> Even if this warning is reasonable, it would be a good idea if the message explained that derive(Debug) is ignored for purposes of the lint. As it happens, I spent a lot of time tearing my hair out trying to figure out why the warning was being triggered. (In my case I have structs that exist only to be dumped to a log using debug prints.)

I think the ideal would be to note that an "ignored impl" is present and issue a `note` on the lint. I'm not sure how easy that would be to do!

### PR DESCRIPTION
**Pull request by FabianWolff**
Fixes #92726. CC @pmetzger, is this what you had in mind?

r? @nikomatsakis

### CODE REVIEW

**Comment by pmetzger** [comment]
Yes, this is more or less what I had in mind! The wording might be tuned very slightly. Right now it seems to say things like 

```
Foo has a derived impl for the trait 'Debug', but this is ignored during dead code analysis
```

It might be slightly more informative to say something to the effect of:

```
Foo does have a derived impl for the trait 'Debug', but this is intentionally ignored for purposes of dead code analysis
```

My suggestion feels like it might be too verbose, but my thought is to make it clear "yes, we know you have a derived impl of Bar, but we have good reasons not to count that when we're figuring out what's dead." Some wordsmithing help would be good here. I bet some small change will make it much more obvious to the reader.

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -666,7 +679,34 @@ impl<'tcx> DeadVisitor<'tcx> {
...
+                            let is_are = if ign_traits.len() > 1 { "these are" } else { "this is" };
+                            let msg = format!("`{}` has {}derived impl{} for the trait{} {}, but {} ignored during dead code analysis",
+                                              self.tcx.item_name(encl_def_id.to_def_id()),
+                                              article,
+                                              plural_s,
+                                              plural_s,
+                                              traits_str,
+                                              is_are);
+                            if let Some(span) = self.tcx.def_ident_span(encl_def_id) {
+                                err.span_note(span, &msg);
```
Comment (nikomatsakis) [inline compiler/rustc_passes/src/dead.rs:702]: would it be possible to get the span of the `Debug` impl instead?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks nice! I'm wondering if we can improve the output by citing the debug impl itself, perhaps with a message like

```
= note: this Debug impl exists, but Debug impls are intentionally ignored by the dead code analysis
```

File: compiler/rustc_passes/src/dead.rs (hunk 1 above)
Comment (FabianWolff) [reply compiler/rustc_passes/src/dead.rs]: Done, the note now points to the part of the derive macro invocation where the `Clone`/`Debug` impl comes from. I've also added the word "intentionally" to the note message for clarity.

**Review by nikomatsakis — APPROVED** [review]
this looks nice :)

**Comment by pmetzger** [comment]
Hooray!

**Pull request merged**